hw/mcu/atmel/sam3x8/src/hal_uart.c

```diff
@@ -0,0 +1,306 @@
+  hal_uart_tx_done tx_done, hal_uart_rx_char rx_func, void *arg){
+    if (uart >= UART_COUNT || uarts[uart].u_open){
+        return -1;
+    }
+    uarts[uart].u_tx_done = tx_done;
+    uarts[uart].u_rx_func = rx_func;
+    uarts[uart].u_tx_func = tx_func;
+    uarts[uart].u_func_arg = arg;
+    return 0;
+}
+
+
+int hal_usart_config(hal_uart_t *uart, int32_t speed, uint8_t databits, uint8_t stopbits,
+                        enum hal_uart_parity parity, enum hal_uart_flow_ctl flow_ctl){
+    if (uart->u_open){
+        return -1;
+    }
+    
+    uart->options.baudrate = speed;
+    
+    /* Set char length */
+    switch (databits){
+        case 5:
+            uart->options.char_length = US_MR_CHRL_5_BIT;
+        break;
+        case 6:
+            uart->options.char_length = US_MR_CHRL_6_BIT;
+        break;
+        case 7:
+            uart->options.char_length = US_MR_CHRL_7_BIT;
+        break;
+        case 8:
+            uart->options.char_length = US_MR_CHRL_8_BIT;
+        break;
+        case 9:
+            uart->options.char_length = US_MR_MODE9;
+        break;
+        default:
+            /* Unsupported mode */
+            return -1;
+    }
+
+    /* Set parity mode */
+    switch (parity){
+        case HAL_UART_PARITY_NONE:
+            uart->options.parity_type = US_MR_PAR_NO;
+        break;
+        case HAL_UART_PARITY_ODD:
+            uart->options.parity_type = US_MR_PAR_ODD;
+        break;
+        case HAL_UART_PARITY_EVEN:
+            uart->options.parity_type = US_MR_PAR_EVEN;
+        break;
+        default:
+            /* Unsupported mode (there are a few that usart supports that this interface does not) */
+            return -1;
+    }
+
+    /* Set stop bits */
+    switch(stopbits){
+        case 1:
+            uart->options.stop_bits = US_MR_NBSTOP_1_BIT;
+        break;
+        case 2:
+            uart->options.stop_bits = US_MR_NBSTOP_2_BIT;
+        break;
+        default:
+            return -1;
+    }   
+
+    /* Ignore flow control, but throw error on illegal argument */
+    switch (flow_ctl){
+        case HAL_UART_FLOW_CTL_NONE:
+        case HAL_UART_FLOW_CTL_RTS_CTS:
+            break;
+        default:
+            return -1;
+    }
+    uart->u_open = 1;
+    /* clock rate should be something like - sysclk_get_peripheral_hz() */
+    if (!usart_init_rs232(uart->uart, &(uart->options), USART_CLOCK_RATE)){
+        return -1;
+    }
+
+    return 0;
+}
+
+/**
+ * hal uart config
+ *
+ * Applies given configuration to UART.
+ */
+int hal_uart_config(int uart, int32_t speed, uint8_t databits, uint8_t stopbits,
+  enum hal_uart_parity parity, enum hal_uart_flow_ctl flow_ctl){
+      void *uart_ptr = uarts[uart].uart;
+      if (is_usart(uart_ptr)){
+          /* USART configuration */
+          /* Assume USART clock and Board are initialized */
+        hal_usart_config(&uarts[uart], speed, databits, stopbits, parity, flow_ctl);
+      }else{
+          /* UART configuration */

```

Comment:
Does UART not need configuration?

Reply:
I'm working on the UART portions now, I just figured I would focus on getting it working for one driver at a time.